src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/NomisTimeServedLicenceService.kt

```diff
@@ -0,0 +1,138 @@
+      NomisTimeServedLicence(
+        nomsId = request.nomsId,
+        bookingId = request.bookingId,
+        reason = request.reason,
+        prisonCode = request.prisonCode,
+        updatedByCa = staff,
+      ),
+    )
+
+    saveAuditEvent(
+      summary = "Recorded NOMIS licence reason",
+      detail = """
+                Created NOMIS licence record:
+                ID=${licenceRecord.id}, nomsId=${licenceRecord.nomsId}, bookingId=${licenceRecord.bookingId},
+                reason=${licenceRecord.reason}, prisonCode=${licenceRecord.prisonCode}
+      """.trimIndent(),
+      staff = staff,
+      changes = mapOf(
+        "nomsId" to licenceRecord.nomsId,
+        "bookingId" to licenceRecord.bookingId,
+        "reason" to licenceRecord.reason,
+        "prisonCode" to licenceRecord.prisonCode,
+      ),
+    )
+  }
+
+  @Transactional
+  fun updateNomisLicenceReason(nomsId: String, bookingId: Long, request: UpdateNomisLicenceReasonRequest) {
+    val licence = licenceRepository.findByNomsIdAndBookingId(nomsId, bookingId)
+      .orElseThrow { EntityNotFoundException("Reason record with nomsId $nomsId and bookingId $bookingId not found") }
+
+    val staff = getCurrentStaff()
+    val oldReason = licence.reason
+
+    licence.reason = request.reason
+    licence.updatedByCa = staff
+    licence.dateLastUpdated = LocalDateTime.now()
+
+    val updatedLicence = licenceRepository.saveAndFlush(licence)
+
+    saveAuditEvent(
+      summary = "Updated NOMIS licence reason",
+      detail = """
+                Updated NOMIS licence record:
+                ID=${updatedLicence.id}, nomsId=${updatedLicence.nomsId}, bookingId=${updatedLicence.bookingId},
+                oldReason=$oldReason, newReason=${updatedLicence.reason}, prisonCode=${updatedLicence.prisonCode}
+      """.trimIndent(),
+      staff = staff,
+      changes = mapOf(
+        "reason (old)" to oldReason,
+        "reason (new)" to updatedLicence.reason,
+        "nomsId" to updatedLicence.nomsId,
+        "bookingId" to updatedLicence.bookingId,
+        "prisonCode" to updatedLicence.prisonCode,
+      ),
+    )
+  }
+
+  @Transactional(readOnly = true)
+  fun findByNomsIdAndBookingId(nomsId: String, bookingId: Long): NomisLicenceReasonResponse? = licenceRepository.findLicenceReasonByNomsIdAndBookingId(nomsId, bookingId)
+
+  @Transactional
+  fun deleteNomisLicenceReason(nomsId: String, bookingId: Long) {
+    val reasonEntity = licenceRepository.findByNomsIdAndBookingId(nomsId, bookingId)
+      .orElseThrow { EntityNotFoundException("No reason found for nomsId=$nomsId and bookingId=$bookingId") }
+
+    val staff = getCurrentStaff()
+
+    saveAuditEvent(
+      summary = "Deleted NOMIS licence reason",
+      detail = """
+                Deleted NOMIS licence record:
+                ID=${reasonEntity.id}, nomsId=${reasonEntity.nomsId}, bookingId=${reasonEntity.bookingId},
+                reason=${reasonEntity.reason}, prisonCode=${reasonEntity.prisonCode}
```

Comment:
Similar to previous comment above, having the reason here might clog the audit a bit